text/0748-glimmer-component-signature.md

```diff
@@ -0,0 +1,461 @@
+---
+Stage: Proposed
+Start Date: 2021-05-13
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/748
+---
+
+## Summary
+
+In TypeScript, the `@glimmer/component` base class currently has a single `Args` type parameter. This parameter declares the names and types of the arguments the component expects to receive.
+
+This RFC proposes a change to that type parameter to become `Signature`, capturing more complete information about how components can be used in a template, including their expected **arguments**, the **block parameters** they provide, and what type of **element(s)** they apply any received attributes and modifiers to.
+
+This RFC is based in large part on prior work by [@gossi] and on learnings from [Glint].
+
+[@gossi]: https://github.com/gossi
+[Glint]: https://github.com/typed-ember/glint
+
+## Motivation
+
+When a developer goes to invoke a component in a template, there are a variety of questions about its interface they need to be able to answer in order to determine how to use it correctly.
+
+1. What arguments does this component accept?
+   - Which arguments are required and which are optional?
+   - What does each argument do?
+   - What types of values am I expected to pass for a given argument?
+2. How can I nest content within this component?
+   - Does it accept a default block? Named blocks?
+   - In what context will my block content appear in the component's output?
+   - For any blocks I write, what parameters does the component expose to them?
```

Comment:
The current proposal does not seem to address all of these questions. The provided type information does not give a hint about the elements allowed within the block in all cases. E.g. if the component yields within a button element, a developer must only use [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content), which is not [interactive](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#interactive_content).

In most cases this could be derived from the `Element` type. But only the component also applies splat attributes to that element. This may not be desirable in all cases.

Also there is the case, in which a parent element, limits the permitted content of a child element. This is the case for any element, which allows transparent content (e.g. `<del>` and `<ins>`). In that case permitted content can not be derived from that element alone.

Not sure if that must be addressed by this RFC though. Maybe it is enough to explicitly name this limitation in the RFC.

Reply:
> The current proposal does not seem to address all of these questions. 

It does; the additional very interesting point you raise about children is an *additional* point. While we could revise the RFC to make it explicit, the answer is implicit in the semantics of templates and unaffected by this proposal: they are simply functions of HTML. This proposal is not trying to specify a way to guarantee that no invalid HTML can be written. It is simply interested in extending the set of things which can be statically known about it for documentation and type checking purposes.

It might be possible to add in a further layer of analysis to accomplish that kind of analysis at a later point, but it is not a current goal for this doc or even for Glint. And it’s worth more that with this proposal and Glint, we are already doing more than any other JS framework. (The one ecosystem I think has taken a swing at the “allowed children” problem is PureScript, though I don’t know if their exploration in that space ended up broadly used even there. It is a very difficult problem because there is no inherent relationship between parent and child elements in HTML syntactically, only semantically.)